apollo-ios-pagination/Sources/ApolloPagination/GraphQLQueryPager.swift

```diff
@@ -1,342 +1,207 @@
-        throw PaginationError.missingInitialPage
-      }
-      guard let nextPageQuery = nextPageResolver(currentPageInfo),
-            currentPageInfo.canLoadMore
-      else { throw PaginationError.pageHasNoMoreContent }
-      guard activeTask == nil else {
-        throw PaginationError.loadInProgress
-      }
-
-      activeTask = Task {
-        let publisher = CurrentValueSubject<Void, Never>(())
-        await withCheckedContinuation { continuation in
-          let watcher = GraphQLQueryWatcher(client: client, query: nextPageQuery) { [weak self] result in
-            guard let self else { return }
-            Task {
-              await self.onSubsequentFetch(
-                cachePolicy: cachePolicy,
-                result: result,
-                publisher: publisher,
-                query: nextPageQuery
-              )
-            }
-          }
-          nextPageWatchers.append(watcher)
-          publisher.sink(receiveCompletion: { [weak self] _ in
-            continuation.resume(with: .success(()))
-            guard let self else { return }
-            Task { await self.onTaskCancellation() }
-          }, receiveValue: { })
-          .store(in: &subscribers)
-          watcher.refetch(cachePolicy: cachePolicy)
-        }
-      }
-      await activeTask?.value
-    }
-
-    public func subscribe(onUpdate: @MainActor @escaping (Result<Output, Error>) -> Void) -> AnyCancellable {
-      $currentValue.compactMap({ $0 }).sink { result in
-        Task {
-          await onUpdate(result)
-        }
-      }
-    }
-
-    /// Reloads all data, starting at the first query, resetting pagination state.
-    /// - Parameter cachePolicy: Preferred cache policy for first-page fetches. Defaults to `returnCacheDataAndFetch`
-    public func refetch(cachePolicy: CachePolicy = .fetchIgnoringCacheData) {
-      assert(firstPageWatcher != nil, "To create consistent product behaviors, calling `fetch` before calling `refetch` will use cached data while still refreshing.")
-      cancel()
-      fetch(cachePolicy: cachePolicy)
-    }
-
-    public func fetch() {
-      cancel()
-      fetch(cachePolicy: .returnCacheDataAndFetch)
-    }
-
-    /// Cancel any in progress fetching operations and unsubscribe from the store.
-    public func cancel() {
-      nextPageWatchers.forEach { $0.cancel() }
-      nextPageWatchers = []
-      firstPageWatcher?.cancel()
-      firstPageWatcher = nil
-
-      varMap = [:]
-      initialPageResult = nil
-      activeTask?.cancel()
-      activeTask = nil
-      subscribers.forEach { $0.cancel() }
-      subscribers.removeAll()
-    }
+private actor Subscriptions {
+  var subscriptions: Set<AnyCancellable> = []
 
-    /// Whether or not we can load more information based on the current page.
-    public var canLoadNext: Bool {
-      currentPageInfo?.canLoadMore ?? false
-    }
+  func store(subscription: AnyCancellable) {
+    subscriptions.insert(subscription)
+  }
+}
```

Comment:
Is it better for us to have these `Subscriptions` as a seperate `actor` from the main `GraphQLQueryPager.Actor`? I imagine that it allows for subscriptions to be accessed in parallel to loading of pages being kicked off, so that might be better for performance. 

Is there any risk of this causing race conditions though? Could a subsscription possibly not be stored in time to get some results it would expect to get when the `Actor` gets results?

Reply:
> Is there any risk of this causing race conditions though? Could a subsscription possibly not be stored in time to get some results it would expect to get when the `Actor` gets results?


Risk seems minimal here. The subscriptions are represented via `CurrentValueSubject`, so it should receive a result as soon as it can. 


> Is it better for us to have these `Subscriptions` as a seperate `actor` from the main `GraphQLQueryPager.Actor`? I imagine that it allows for subscriptions to be accessed in parallel to loading of pages being kicked off, so that might be better for performance.

The `Subscriptions` actor only exists to manage the subscriptions array in `GraphQLQueryPager`. if we omit this `actor`, the alternative would be to store individual values for each `AnyCancellable` on the `GraphQLQueryPager`. Otherwise, we risk allowing concurrent access of the `subscriptions` array/set, which can lead to a crash.